### PR TITLE
Refactor the anchor in SiameseRPN++

### PR DESCRIPTION
**Pull request by JingweiZhang12**
This PR mainly refactors the anchor in SiameseRPN++ as follows :
1. converts the anchor format from `[cx, cy, w, h]` to `[tl_x, tl_y, br_x, br_y]`.
2. uses new anchor order in order to be consistent with MMDet.
3. replaces the `grid_anchors` with `grid_priors` to skip the deprecated warnings in MMDet.

### CODE REVIEW

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 1]
```diff
@@ -113,18 +114,24 @@ def single_level_grid_anchors(self,
...
+        # then (0, 1), (0, 2), ...
         all_anchors = base_anchors[:, None, :] + shifts[None, :, :]
         all_anchors = all_anchors.view(-1, 4)
 
+        # transform the coordinate origin from the top left corner to the
+        # center in the scaled featurs map.
         all_anchors[:, 0] += -(feat_w // 2) * stride[0]
         all_anchors[:, 1] += -(feat_h // 2) * stride[1]
+        all_anchors[:, 2] += -(feat_w // 2) * stride[0]
+        all_anchors[:, 3] += -(feat_h // 2) * stride[1]
```
Comment (GT9505) [inline mmtrack/core/anchor/sot_anchor_generator.py:135]: we can simplify it by using slices.
`all_anchors[:, 0:4:2] += -(feat_w // 2) * stride[0]`

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 2]
```diff
@@ -113,18 +114,24 @@ def single_level_grid_anchors(self,
...
+        # first feat_w elements correspond to the first row of shifts
+        # add A anchors (1, A, 4) to K shifts (K, 1, 4) to get
+        # shifted anchors (K, A, 4), reshape to (K*A, 4)
+        # all_anchors = base_anchors[None, :, :] + shifts[:, None, :]
+        # first A rows correspond to A anchors of (0, 0) in feature map,
+        # then (0, 1), (0, 2), ...
         all_anchors = base_anchors[:, None, :] + shifts[None, :, :]
         all_anchors = all_anchors.view(-1, 4)
 
+        # transform the coordinate origin from the top left corner to the
```
Comment (GT9505) [inline mmtrack/core/anchor/sot_anchor_generator.py:130]: Please capitalize the first letter. Please also fix the problem in other comments。

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 3]
```diff
@@ -55,7 +55,7 @@ def gen_single_level_base_anchors(self,
 
         Returns:
             torch.Tensor: Anchors of one spatial location in a single level
-            feature map in [cx, cy, w, h] format.
+            feature map in [tl_x, tl_y, br_x, br_y] format.
```
Comment (GT9505) [inline mmtrack/core/anchor/sot_anchor_generator.py:58]: This function is the same as function in mmdet. We can directly delete it.

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 4]
```diff
@@ -113,18 +114,24 @@ def single_level_grid_anchors(self,
...
+        shifts = torch.stack([shift_xx, shift_yy, shift_xx, shift_yy], dim=-1)
         shifts = shifts.type_as(base_anchors)
 
+        # TODO
+        # first feat_w elements correspond to the first row of shifts
+        # add A anchors (1, A, 4) to K shifts (K, 1, 4) to get
+        # shifted anchors (K, A, 4), reshape to (K*A, 4)
+        # all_anchors = base_anchors[None, :, :] + shifts[:, None, :]
+        # first A rows correspond to A anchors of (0, 0) in feature map,
+        # then (0, 1), (0, 2), ...
```
Comment (GT9505) [inline mmtrack/core/anchor/sot_anchor_generator.py:126]: We can fix the `TODO` in the pr in order to be the same with mmdet

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 5]
```diff
@@ -436,8 +439,7 @@ def get_bbox(self, cls_score, bbox_pred, prev_bbox, scale_factor):
 
         bbox_pred = bbox_pred.permute(1, 2, 3, 0).contiguous().view(4, -1)
         bbox_pred = bbox_pred.permute(1, 0)
-        anchors = bbox_cxcywh_to_xyxy(self.anchors)
-        bbox_pred = self.bbox_coder.decode(anchors, bbox_pred)
+        bbox_pred = self.bbox_coder.decode(self.anchors, bbox_pred)
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:442]: Be careful about the `permute` operation, since we changed the order of anchors. We need check the `permute` operation in `get_bbox` and `loss` functions in order to keep them consistent.

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 6]
```diff
@@ -38,93 +38,43 @@ def gen_2d_hanning_windows(self, featmap_sizes, device='cuda'):
...
 
-        all_anchors[:, 0] += -(feat_w // 2) * stride[0]
-        all_anchors[:, 1] += -(feat_h // 2) * stride[1]
-
+        # Transform the coordinate origin from the top left corner to the
+        # center in the scaled featurs map.
+        all_anchors[:, 0:4:2] += -(feat_w // 2) * stride_w
+        all_anchors[:, 1:4:2] += -(feat_h // 2) * stride_h
```
Comment (GT9505) [inline mmtrack/core/anchor/sot_anchor_generator.py:79]: We can delete this function and move line 78-79 outside of the function, since the function is almost the same as the function in mmdet

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 7]
```diff
@@ -197,17 +197,15 @@ def forward(self, z_feats, x_feats):
...
-        bbox_targets = torch.zeros((4, num_base_anchors, score_maps_size[0],
-                                    score_maps_size[1])).to(gt_bbox.device)
-        bbox_weights = torch.zeros((num_base_anchors, score_maps_size[0],
-                                    score_maps_size[1])).to(gt_bbox.device)
+        H, W = score_maps_size
+        num_anchors = num_base_anchors * H * W
+        labels = gt_bbox.new_zeros((num_anchors, ), dtype=torch.long)
+        labels_weights = gt_bbox.new_zeros((num_anchors, ), dtype=torch.float)
+        bbox_weights = gt_bbox.new_zeros((num_anchors, ), dtype=torch.float)
+        bbox_targets = gt_bbox.new_zeros((num_anchors, ), dtype=torch.float)
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:208]: `bbox_targets = gt_bbox.new_zeros((num_anchors, 4), dtype=gt_bbox.dtype)`

**Comment by GT9505** [comment]
The unittest failed.

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 8]
```diff
@@ -353,9 +348,9 @@ def get_targets(self, gt_bboxes, score_maps_size, is_positive_pairs):
 
         Returns:
             tuple(all_labels, all_labels_weights, all_bbox_targets,
-            all_bbox_weights): the shape is (N, num_base_anchors, H, W),
-            (N, num_base_anchors, H, W), (N, 4, num_base_anchors, H, W),
-            (N, 4, num_base_anchors, H, W), respectively. All of them are
+            all_bbox_weights): the shape is (N, H * W * num_base_anchors),
+            (N, H * W * num_base_anchors), (N, H * W * num_base_anchors, 4),
+            (N, H * W * num_base_anchors), respectively. All of them are
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:353]: the shape of last tensor is `(N, H * W * num_base_anchors, 4)`

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 9]
```diff
@@ -394,10 +389,10 @@ def loss(self, cls_score, bbox_pred, labels, labels_weights, bbox_targets,
...
             cls_score (Tensor): of shape (N, 2 * num_base_anchors, H, W).
             bbox_pred (Tensor): of shape (N, 4 * num_base_anchors, H, W).
-            labels (Tensor): of shape (N, num_base_anchors, H, W).
-            labels_weights (Tensor): of shape (N, num_base_anchors, H, W).
-            bbox_targets (Tensor): of shape (N, 4, num_base_anchors, H, W).
-            bbox_weights (Tensor): of shape (N, 4, num_base_anchors, H, W).
+            labels (Tensor): of shape (N, H * W * num_base_anchors).
+            labels_weights (Tensor): of shape (N, H * W * num_base_anchors).
+            bbox_targets (Tensor): of shape (N, H * W * num_base_anchors, 4).
+            bbox_weights (Tensor): of shape (N, H * W * num_base_anchors).
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:395]: `bbox_weights (Tensor): of shape (N, H * W * num_base_anchors, 4).`

File: mmtrack/core/anchor/sot_anchor_generator.py [hunk 10]
```diff
@@ -34,7 +34,7 @@ def gen_2d_hanning_windows(self, featmap_sizes, device='cuda'):
             hanning_h = np.hanning(featmap_sizes[i][0])
             hanning_w = np.hanning(featmap_sizes[i][1])
             window = np.outer(hanning_h, hanning_w)
-            window = np.tile(window.flatten(), self.num_base_anchors[i])
+            window = window.flatten().repeat(self.num_base_anchors[i])
```
[the remark this answers is not in the document]
Comment (JingweiZhang12) [reply mmtrack/core/anchor/sot_anchor_generator.py:37]: Yes, I have checked it

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 11]
```diff
@@ -137,8 +137,8 @@ def __init__(self,
                  init_cfg=None,
                  *args,
                  **kwargs):
-        super(SiameseRPNHead, self).__init__(init_cfg)
-        self.anchor_generator = build_anchor_generator(anchor_generator)
+        super(SiameseRPNHead, self).__init__(*args, **kwargs)
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:140]: We need to pass `init_cfg` into the super class.

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 12]
```diff
@@ -207,17 +207,15 @@ def forward(self, z_feats, x_feats):
...
-                                      score_maps_size[1])).to(gt_bbox.device)
-        bbox_targets = torch.zeros((4, num_base_anchors, score_maps_size[0],
-                                    score_maps_size[1])).to(gt_bbox.device)
-        bbox_weights = torch.zeros((num_base_anchors, score_maps_size[0],
-                                    score_maps_size[1])).to(gt_bbox.device)
+        H, W = score_maps_size
+        num_anchors = H * W * num_base_anchors
+        labels = gt_bbox.new_zeros((num_anchors, ), dtype=torch.long)
+        labels_weights = gt_bbox.new_zeros((num_anchors, ))
+        bbox_weights = gt_bbox.new_zeros((num_anchors, ))
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:217]: `bbox_weights = gt_bbox.new_zeros((num_anchors, 4))`

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 13]
```diff
@@ -272,12 +275,7 @@ def _get_positive_pair_targets(self, gt_bbox, score_maps_size):
 
         bbox_targets = self.bbox_coder.encode(
             anchors, gt_bbox[:, 1:].repeat(anchors.shape[0], 1))
-
-        labels = labels.reshape(C, H, W)
-        labels_weights = labels_weights.reshape(C, H, W)
-        bbox_targets = bbox_targets.T.reshape(4, C, H, W)
-        bbox_weights = bbox_weights.reshape(C, H, W)
-        bbox_weights = bbox_weights[None].repeat(4, 1, 1, 1)
+        bbox_weights = bbox_weights.view(-1, 1).repeat(1, 4)
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:278]: delete this line

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 14]
```diff
@@ -436,20 +434,27 @@ def get_bbox(self, cls_score, bbox_pred, prev_bbox, scale_factor):
         """
         score_maps_size = [(cls_score.shape[2:])]
         if not hasattr(self, 'anchors'):
-            self.anchors = self.anchor_generator.grid_anchors(
+            self.anchors = self.anchor_generator.grid_priors(
                 score_maps_size, cls_score.device)[0]
+            # Transform the coordinate origin from the top left corner to the
+            # center in the scaled featurs map.
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:440]: `feature map`

File: mmtrack/models/track_heads/siamese_rpn_head.py [hunk 15]
```diff
@@ -333,9 +330,7 @@ def _get_negative_pair_targets(self, gt_bbox, score_maps_size):
             labels_weights[neg_inds] = 1.0 / len(neg_inds) / 2
         labels[...] = 0
 
-        labels = labels.reshape(C, H, W)
-        labels_weights = labels_weights.reshape(C, H, W)
-        bbox_weights = bbox_weights[None].repeat(4, 1, 1, 1)
+        bbox_weights = bbox_weights.view(-1, 1).repeat(1, 4)
```
Comment (GT9505) [inline mmtrack/models/track_heads/siamese_rpn_head.py:333]: delete this line

**Review by GT9505 — APPROVED** [review]
